### DISCUSSION TITLE
Possible Port to STM32/Daisy

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/arduino-stk#9] [Q&A] [answered]
I am interested in extending this port to work with the Daisy (stm32H7)

There are no "streams" per se on the daisy. A high priority Audio Callback which passes to the i2s Codec is always available and there is a built in Tick function which counts per sample.

My initial thought was that I would need to rewrite ArdStreamOut.h but it is possible I could just take care of that in the audio callback itself.

below is a typical callback in Daisy

`void AudioCallback(AudioHandle::InputBuffer in, AudioHandle::OutputBuffer out, size_t size)
{
	float coarse_knob = hw.adc.GetFloat(pot1); 
    float freq = fmap(coarse_knob, 0.f, 1000.f);

	float detune_knob = hw.adc.GetFloat(pot2); 

	osc_a.SetFreq(freq); 
	osc_b.SetFreq(freq + (.05 * freq * detune_knob)); 
	osc_c.SetFreq(freq - (.05 * freq * detune_knob)); 

	for (size_t i = 0; i < size; i++)
	{
		float output = osc_a.Process() + osc_b.Process() + osc_c.Process(); 

		out[0][i] = output;
		out[1][i] = output;
	}
}`

**Comment by pschatzmann** [comment, marked as answer]
You did not specify what development environement you want to use, so I assume that you want to use DaisyDuino.
If so, check out the [following discussion](https://github.com/pschatzmann/arduino-audio-tools/discussions/992)

The far bigger challenge that I see however, is that I replaced the Posix file system commands with a mechanism that takes the sound files from Program Memory and on a STM you will just not have enough! 

So you you will need to deactivate this in the configuration file and somehow make the file operations work in your environment e.g with an SD card.

**Reply by SnazzyBlanket** [reply]
Oh sorry No i want to use VS CODE

**Reply by SnazzyBlanket** [reply]
the daisy has 64mb of onboard memory-- I would need to figure out how to load it into that storage area

**Reply by SnazzyBlanket** [reply]
I think it would be as easy as defining those variables as 
type float DSY_SDRAM_BSS

**Reply by SnazzyBlanket** [reply]
I found your link

https://github.com/pschatzmann/arduino-audio-tools/wiki/Writing-your-Custom-Input&Output-Class

However it is this Arduino specific? (not sure if audiotools.h calls any wire.h specific functions etc. 

Also from a conceptual standpoint I am trying to come to terms with how the STK deals with sound. if it is a matter of passing a tick to the sine function and then that function returns a value that can be output to the callback then all i would need to do is pass the tick. But I have a feeling there is more going on? thanks for your help

**Comment by pschatzmann** [comment]
I am sorry, your statement about VS Code does not make any sense. This is a Programming Editor and does not define which API to the processor you will use: You can even do your Arduino (or in your case DaisyDuino) Sketches there.

I tried to look at the Daisy Spec but for some strange reason there is no information of how much PROGMEM you have available or about what file API to expect. I guess you need to figure this one out yourself.

Though the AudioTools have been written primarily for Arduino, but there is a small compatibility layer that lets it run in any other environment: That's how you can compile your audio sketches on the Desktop! Further information you can find in the [Wiki of the AudioTools Project](https://github.com/pschatzmann/arduino-audio-tools/wiki).

ps.
But you are always free to implement your own output class directly: This would write the audio data to a queue and your callback would get it back from there. Just have a look at ArdCommonOut.h for the first step.

**Reply by SnazzyBlanket** [reply]
Sorry--- Though VSCode is just a programming Editor, they have set it up in such a way where they have setup everything in such a way where most users either choose the DaisyDuino framework or Cmake/Make in VsCode.

I will check out the link you posted and will look at the ArdCommonOut.h

**Reply by SnazzyBlanket** [reply]
thanks for your help

**Reply by SnazzyBlanket** [reply]
If I can get it going I will post about it here

**Reply by SnazzyBlanket** [reply]
and just a note..there are definitely some aspects of the Daisy Documentation which are lacking or hard to find etc.

anyway here is the best link

https://daisy.audio/libDaisy/Pages/md_doc_md__a6__getting__started__external__s_d_r_a_m/#page-md-doc-md--a6-getting-started-external-sdram

**Comment by pschatzmann** [comment]
Taking into consideration that both the Daisy API and arduino-stk use floats, you should prefer to write your own small output class based on ArdCommonOut.h
